src/query/users/src/visibility_checker.rs

```diff
@@ -54,12 +61,52 @@ impl GrantObjectVisibilityChecker {
         for role in available_roles {
             grant_sets.push(&role.grants);
         }
-
         for grant_set in grant_sets {
             for ent in grant_set.entries() {
                 match ent.object() {
                     GrantObject::Global => {
-                        granted_global = true;
+                        fn check_privilege<T, F>(granted: &mut bool, mut iter: T, condition: F)
+                        where
+                            T: Iterator<Item = UserPrivilegeType>,
+                            F: Fn(UserPrivilegeType) -> bool,
+                        {
+                            if !*granted {
+                                *granted |= iter.any(condition);
+                            }
+                        }
+
+                        check_privilege(
+                            &mut granted_global_udf,
+                            ent.privileges().iter(),
+                            |privilege| {
+                                UserPrivilegeSet::available_privileges_on_udf(false)
+                                    .has_privilege(privilege)
+                            },
+                        );
+
+                        check_privilege(
+                            &mut granted_global_stage,
+                            ent.privileges().iter(),
+                            |privilege| {
+                                UserPrivilegeSet::available_privileges_on_stage(false)
+                                    .has_privilege(privilege)
+                            },
+                        );
+
+                        check_privilege(
+                            &mut granted_global_read_stage,
+                            ent.privileges().iter(),
+                            |privilege| privilege == UserPrivilegeType::Read,
+                        );
+
+                        check_privilege(
+                            &mut granted_global_db_table,
+                            ent.privileges().iter(),
+                            |privilege| {
+                                UserPrivilegeSet::available_privileges_on_database(false)
+                                    .has_privilege(privilege)
+                            },
```

Comment:
is this check validates every granted entry's privileges are contained in global privilege available on database?

Reply:
Yes. But only global object will in this match arm

Reply:
I suppose that the checking logic is not very straight forward for new comers, how about add a comment above this check?

Reply:
Added.